components/api/src/main/kotlin/com/hotels/styx/api/MeterRegistry.kt

```diff
@@ -0,0 +1,216 @@
+        override fun longTaskTimer(name: String, tags: Iterable<Tag>): LongTaskTimer = more.longTaskTimer(name.scoped, tags)
+
+        override fun <T> counter(name: String, tags: Iterable<Tag>, obj: T, countFunction: ToDoubleFunction<T>): FunctionCounter =
+            more.counter(name.scoped, tags, obj, countFunction)
+
+        override fun <T : Number> counter(name: String, tags: Iterable<Tag>, number: T): FunctionCounter =
+            more.counter(name.scoped, tags, number)
+
+        override fun <T> timer(
+            name: String,
+            tags: Iterable<Tag>,
+            obj: T,
+            countFunction: ToLongFunction<T>,
+            totalTimeFunction: ToDoubleFunction<T>,
+            totalTimeFunctionUnit: TimeUnit
+        ): FunctionTimer = more.timer(name.scoped, tags, obj, countFunction, totalTimeFunction, totalTimeFunctionUnit)
+
+        override fun <T> timeGauge(
+            name: String,
+            tags: Iterable<Tag>,
+            obj: T,
+            timeFunctionUnit: TimeUnit,
+            timeFunction: ToDoubleFunction<T>
+        ): TimeGauge = more.timeGauge(name.scoped, tags, obj, timeFunctionUnit, timeFunction)
+    }
+
+    override fun <T : Number> gauge(name: String, number: T): T = registry.gauge(name.scoped, number)
+    override fun <T> gauge(name: String, stateObject: T, valueFunction: ToDoubleFunction<T>): T =
+        registry.gauge(name.scoped, stateObject, valueFunction)
+
+    override fun <T> gauge(name: String, tags: Iterable<Tag>, stateObject: T, valueFunction: ToDoubleFunction<T>): T =
+        registry.gauge(name.scoped, tags, stateObject, valueFunction)
+
+    override fun <T : Number> gauge(name: String, tags: Iterable<Tag>, number: T): T = registry.gauge(name.scoped, tags, number)
+
+    override fun <T : Collection<*>> gaugeCollectionSize(name: String, tags: Iterable<Tag>, collection: T): T =
+        registry.gaugeCollectionSize(name.scoped, tags, collection)
+
+    override fun <T : Map<*, *>> gaugeMapSize(name: String, tags: Iterable<Tag>, map: T): T = registry.gaugeMapSize(name.scoped, tags, map)
+
+    override fun remove(meter: Meter): Meter? = registry.remove(meter)
+    override fun remove(mappedId: Meter.Id): Meter? = registry.remove(mappedId.scoped)
+    override fun clear() = registry.clear()
+    override fun close() = registry.close()
+    override fun isClosed() = registry.isClosed()
+
+    // Functions below here are not for general metric handling, but needed by Styx internally.
+
+    override fun startTimer(): Timer.Sample = registry.startTimer()
+    override fun micrometerRegistry(): io.micrometer.core.instrument.MeterRegistry? = registry.micrometerRegistry()
+
+    private val String.scoped: String get() = "$scope.$this"
+    private val Meter.Id.scoped: Meter.Id get() = withName(name.scoped)
+}
+
+class MicrometerRegistry(private val registry: io.micrometer.core.instrument.MeterRegistry) : MeterRegistry {
```

Comment:
If above class took a Micrometer MeterRegistry, this class would be obsolete

Reply:
This is incorrect. If we did that we would only be able to have a single level of scoping. The current implementation allows `ScopedMeterRegistry` to wrap other instances of itself, allowing for multi-level scoping.

Reply:
ah ok, yes.  You explained this to me already.  I forgot 🤦